45_api_versioning.md

```diff
@@ -0,0 +1,162 @@
+
+#### Story 3
+
+An organization has upgraded their machine images to a new version of the `keylime-agent` which uses the v1.5 version of the API. Their existing machines use a previous version of the agent which uses the v1.2 version. Their server-side components are currently supporting v1.4 of the API. They first upgrade their server-side components to the latest version which is now at v2.1 of the API. Then new machines with the v1.5 client can communicate with the server side and existing clients running v1.4 will also continue to operate without problems. The fleet should be able to operate in this "mixed" mode of different client versions without issue.
+
+### Notes/Constraints/Caveats
+
+* The currect API will be pegged at `v1.0` and when the feature is released all components will be updated together. This will be the last time that this will be required as future versions will allow the client and server-side components to upgrade independently. But we need this clean break from previous URL paths to give us this foundation for the future.
+* It's difficult to predict the future, so it may not be possible to restrict the project to supporting just the latest 2 major versions, but that's a good place to start.
+* This proposal does not include any API version negotiation. It should be sufficient for the client code to use a single version as long as it's in a major version family that the server-side supports.
+* This proposal does not guarantee the ability for a newer client (using say `v3.4`) to communicate with an older server-side component (using say `v3.3`). Backwards compatible changes are within scope, not future-compatible ones. But it also doesn't explicitly disallow it. In some cases it might work out in practice for scenarios like the following: The only difference between `v3.4` and `v3.3` is the addition of a new optional field. The server side would just ignore this new unknown field. But any organization doing this would have no guarantees and would need to do their own extensive vetting.
+
+
+### Risks and Mitigations
+
+There is always the risk that this proposal is not flexible enough and that we would need to revist in the future. Hopefully this can be done without needing another "clean break" in API URL versioning and can just be built on top of this one with new strictures and guarantees.
+
+It's possible that there are security vulnerabilities discovered in a previous version that is still supported where those vulnerabilities are inherent in how the API is structured. If such a vulnerability could not be fixed without breaking the API then we would need to review as a project how to deprecate support for that particular version without breaking the whole major version family if possible. If that is not possible, some other mitigation strategies would need to be decided on a case-by-case basis.
```

Comment:
This is a good point. I was also wondering to what extent this enhancement should concern itself with possible downgrade attacks, if an earlier version of the API is found to have a weakness that can be exploited. What do you think?

Reply:
Yeah, it's something we should be aware of when we fix security vulnerabilities, but I'm not sure we can solve it globally in the proposal. I think we'll need to handle each one on a case-by-case basis.

Reply:
My concern is at a higher level.  

Currently the verifier does a GET to the agent, with parameters encoded in the URI.  If I have it right, the length must be less than 2k, and that includes the FQDN.  As soon as keylime wants to send anything larger (like a certificate or something else we haven't thought of), the GET scheme breaks.

If keylime is doing a one time rework, can the 'GET with URI encoded parameters' be replaced with a more flexible json packet?

Does the registrar have a similar fixed small limit on the packet size?

Reply:
This purpose of this proposal is to give us a framework to make future changes like the one you're suggesting. But we won't be doing those changes as part of this. I suspect that changes like the one you're suggesting are going to be some of the first we'll do as part of v2.0 of the APIs.